relay/shfu.go

```diff
@@ -0,0 +1,325 @@
+}
+
+// getUpdateClientResultsFromGRPC retrieves SHFU results from gRPC server using height range
+func getUpdateClientResultsFromGRPC(ctx context.Context, logger *log.RelayLogger, grpcAddress string, targetChain core.Chain, counterparty core.FinalityAwareChain, latestFinalizedHeader core.Header) ([]*shfu_storage.UpdateClientResult, error) {
+	logger.InfoContext(ctx, "using SHFU gRPC server", "address", grpcAddress)
+
+	// Get chain ID from target chain and counterparty chain
+	chainID := targetChain.ChainID()
+	counterpartyChainID := counterparty.ChainID()
+
+	// Get SHFU record by height range
+	counterpartyLatestFinalizedHeader, err := counterparty.GetLatestFinalizedHeader(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest finalized header: %w", err)
+	}
+	fromHeight, err := GetClientStateHeight(ctx, counterparty, counterpartyLatestFinalizedHeader.GetHeight())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get client state height from counterparty chain: %w", err)
+	}
+
+	records, err := shfu_grpc.GetSequentialSHFURecords(ctx, grpcAddress, chainID, counterpartyChainID, fromHeight, latestFinalizedHeader.GetHeight())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sequential SHFU records: %w", err)
+	}
+
+	var results []*shfu_storage.UpdateClientResult
+	var heights []string
+	for _, record := range records {
+		heights = append(heights, fmt.Sprintf("%d-%d", record.ToHeight.RevisionNumber, record.ToHeight.RevisionHeight))
+		results = append(results, record.UpdateClientResults...)
+	}
+	logger.InfoContext(ctx, "retrieved SHFU records from gRPC server",
+		"chain_id", chainID,
+		"counterparty_chain_id", counterpartyChainID,
+		"heights", strings.Join(heights, ", "),
+	)
+
+	return results, nil
+}
+
+func getTipHeightInStorage(ctx context.Context, targetChainID string, counterpartyChainID string, fromHeight ibcexported.Height, storage SHFUStorage, logger *log.RelayLogger) (ibcexported.Height, error) {
+	records, err := storage.GetSequentialSHFURecords(ctx, targetChainID, counterpartyChainID, fromHeight, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sequential SHFU records: %w", err)
+	}
+
+	if len(records) == 0 {
+		return nil, nil
+	}
+
+	// Log the sequential records found
+	if logger != nil {
+		// Create comma-separated height list for debugging
+		var heightList []string
+		for _, record := range records {
+			heightList = append(heightList, fmt.Sprintf("%s..%s", record.FromHeight.String(), record.ToHeight.String()))
+		}
+		heightListStr := strings.Join(heightList, ",")
+
+		logger.InfoContext(ctx, "Found sequential SHFU records",
+			"chain_id", targetChainID,
+			"counterparty_chain_id", counterpartyChainID,
+			"starting_height", fromHeight.String(),
+			"records_count", len(records),
+			"height_list", heightListStr)
+	}
+
+	return records[len(records)-1].ToHeight, nil
+}
+
+// ExecuteSetupHeadersForUpdate executes updateELCForUpdateClient and returns UpdateClientResult array
```

Comment:
The first word should match the function name:

```suggestion
// SHFUExecuteAndStore executes updateELCForUpdateClient and returns UpdateClientResult array
```

Reply:
fixed